### DISCUSSION TITLE
Dynamic header on a static page

### DISCUSSION THREAD

**Discussion by nedak96** [vercel/next.js#51187] [App Router]
Hello! I'm curious if there is any guidance on how to go about having some dynamic content based on auth state, but otherwise having static content. Here is a basic example

### layout.tsx
```jsx
import { cookies } from 'next/headers';

const Header = () => {
  const isLoggedIn = cookies().get('access-token');

  return (
    <header>
      {isLoggedIn ? <button>Logout</button> : <button>Login</button>}
    </header>
  );
};

const Layout = ({ children }) => {
  return (
    <htm>
      <body>
        <Header />
        <main>{children}</main>
      </body>
    </html>
  );
};

export default Layout;
```

### page.tsx

(I want to statically generate this content)
```jsx
import { prisma } from 'lib/prisma';

const Home = async () => {
  const items = await prisma.items.findMany();

  return (
    <ul>
      {items.map((item) => <li>{item.name}</li>}
    </ul>
  );
};

return default Home;
```

When I build this page, I get the message `warn Entire page / deopted into client-side rendering` and the page refetches the data on every load and SSR the HTML. I know the dynamic header would need to be SSR'd, but I'm less worried about that.

One solution that seems to work is separating these into two pages and using middleware to rewrite based on auth state

###  middleware.ts
```typescript
const middleware = async (req) => {
  const isLoggedIn = !!req.cookies.get('access-token')?.value;
  if (req.nextUrl.pathname === '/' && isLoggedIn) {
    const url = req.nextUrl.clone();
    url.pathname = '/protected';
    return NextResponse.rewrite(url);
  }
  if (req.nextUrl.pathname === '/items' && !isLoggedIn) {
    const url = req.nextUrl.clone();
    url.pathname = '/';
    return NextResponse.redirect(url);
  }
  return NextResponse.next();
}

return default middleware;
```

### layout.tsx
```jsx
const Layout = ({ children }) => {
  return (
    <htm>
      <body>
        {children}
      </body>
    </html>
  );
};

export default Layout;
```

###  (authenticatedRoutes)/layout.tsx
```jsx
const AuthenticatedLayout = ({ children }) => {
  return (
    <>
      <header>
        <button>Logout</button>
      </header>
      <main>{children}</main>
    </>
  );
};

export default AuthenticatedLayout;
```

###  (authenticatedRoutes)/protected/page.tsx
```jsx
import { prisma } from 'lib/prisma';

const Home = async () => {
  const items = await prisma.items.findMany();

  return (
    <ul>
      {items.map((item) => <li>{item.name}</li>}
    </ul>
  );
};

return default Home;
```

###  (unauthenticatedRoutes)/layout.tsx
```jsx
const UnauthenticatedLayout = ({ children }) => {
  return (
    <>
      <header>
        <button>Login</button>
      </header>
      <main>{children}</main>
    </>
  );
};

export default UnauthenticatedLayout;
```

###  (unauthenticatedRoutes)/page.tsx
```jsx
import { prisma } from 'lib/prisma';

const Home = async () => {
  const items = await prisma.items.findMany();

  return (
    <ul>
      {items.map((item) => <li>{item.name}</li>}
    </ul>
  );
};

return default Home;
```

This seems to work, but there is much more code than I'd like there to be. Does anyone have guidance on the best way to do this? Maybe just [caching the db response](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#conditional-routes) is enough, or maybe even [conditional parallel routes](https://nextjs.org/docs/app/building-your-application/routing/parallel-routes#conditional-routes). Thanks in advance!

**Comment by mxhold** [comment]
[Partial Prerendering](https://rc.nextjs.org/docs/app/api-reference/next-config-js/ppr) might help you out with this though it is currently experimental.
